src/main/java/Sorting.java

```diff
@@ -0,0 +1,273 @@
+
+        if(start < end) {
+            int middle = (int)Math.floor((start + end) / 2d);
+            mergeSort(array, start, middle);
+            mergeSort(array, middle + 1, end);
+            merge(array,start,middle,end);
+        }
+
+        return  array;
+    }
+
+    private static void merge(List<Integer> array, int start, int middle, int end) {
+
+        // heapArray.subList(inclusive,exclusive) so middle will be start of right heapArray
+        middle++;
+        // heapArray.subList(start,end) will throw exception if(start < 0 || end > size)
+        end++;
+        LinkedList<Integer> leftArray = new LinkedList<Integer>(array.subList(start, middle));
+        LinkedList<Integer> rightArray = new LinkedList<Integer>(array.subList(middle, end));
+
+        int index = start;
+        // two possible problems areas here: arrays are emtpy before or after end
+        while(! (leftArray.isEmpty() && rightArray.isEmpty())){
+
+            int value;
+            if(leftArray.isEmpty()) {
+                value = rightArray.remove();
+            } else if(rightArray.isEmpty()) {
+                value = leftArray.remove();
+            } else {
+                if(leftArray.getFirst() < rightArray.getFirst()) {
+                    value = leftArray.remove();
+                } else {
+                    value = rightArray.remove();
+                }
+            }
+            array.set(index, value);
+            index++;
+        }
+    }
+
+
+    // implement heap sort
+    /*
+     * heap is weak binary tree stored as an array list
+     * max heap has max at root, min heap has min at root
+     * parent node of node at index k is Math.round((k-1)/2)
+     * left child node is 2k + 1
+     * right child node is left child node + 1
+     * delete swaps top and bottom nodes, deletes bottom node, sifts down top node
+     * sift down will compare both children for max (or min in min heap) and then if max > node, swap
+     * sift up will compare to parent, if greater, swap
+     * method swap
+     */
+
+    public static List<Integer> heapSort(List<Integer> array)  throws IllegalArgumentException {
+
+        if(array == null) {
+            throw new IllegalArgumentException("invalid input array to heap sort: input array is null");
+        }
+
+        if(array.isEmpty()) {
+            throw new IllegalArgumentException("invalid input array to heap sort: empty array");
+        }
+
+        LinkedList<Integer> list = new LinkedList<Integer>(array);
+        // I guess under the hood a PriorityQueue is a min heap
+        // and to do a max heap, the PQ can be made with a comparator?
+        // but this implementation doesnt involve implementing the heap itself
+        // just knowing the PQ behaves like a heap with poll and add
+        PriorityQueue<Integer> heapArray = new PriorityQueue<Integer>(10);
+        while(! list.isEmpty()) {
+            heapArray.add(list.poll());
```

Comment:
just iterate list and add each element to heapArray, then create a new List, pop each element there, and return it. same comment about not touching inputs to the function (unless clearly noted in the Javadoc, like Collections.swap)